File: app/src/main/java/com/cdiamon/autocolorist/tables/DBHelper.java
package com.cdiamon.autocolorist.tables;

import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Created by Dmitriy on 28.12.2016.
 */

class DBHelper extends SQLiteOpenHelper {


    //The Android's default system path of your application database.

    private static final String DATABASE_PATH = "/data/data/com.cdiamon.autocolorist/databases/";
    private static final String DATABASE_NAME = "converttables.db";
    private static final String DATABASE_TABLE = "paints_name";
    private static final int DATABASE_VERSION = 1;

    private SQLiteDatabase myDataBase;
    private final Context myContext;

    String QueryString = "";

    final String[] componentNames = new String[30];

    /**
     * Constructor
     * Takes and keeps a reference of the passed context in order to access to the application assets and resources.
     *
     * @param context
     */

    DBHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        this.myContext = context;
    }

    /**
     * Creates a empty database on the system and rewrites it with your own database.
     */
    void createDataBase() throws IOException {

        boolean dbExist = checkDataBase();

        if (dbExist) {
            //do nothing - database already exist
        } else {

            //By calling this method and empty database will be created into the default system path
            //of your application so we are gonna be able to overwrite that database with our database.
            this.getReadableDatabase();

            try {

                copyDataBase();

            } catch (IOException e) {

                throw new Error("Error copying database");

            }
        }


    }

    /**
     * Check if the database already exist to avoid re-copying the file each time you open the application.
     *
     * @return true if it exists, false if it doesn't
     */
    private boolean checkDataBase() {

        SQLiteDatabase checkDB = null;

        try {
            String myPath = DATABASE_PATH + DATABASE_NAME;
            checkDB = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY);

        } catch (SQLiteException e) {

            //database does't exist yet.

        }

        if (checkDB != null) {

            checkDB.close();

        }

        return checkDB != null ? true : false;
    }

    /**
     * Copies your database from your local assets-folder to the just created empty database in the
     * system folder, from where it can be accessed and handled.
     * This is done by transfering bytestream.
     */
    private void copyDataBase() throws IOException {

        //Open your local db as the input stream
        InputStream myInput = myContext.getAssets().open(DATABASE_NAME);

        // Path to the just created empty db
        String outFileName = DATABASE_PATH + DATABASE_NAME;

        //Open the empty db as the output stream
        OutputStream myOutput = new FileOutputStream(outFileName);

        //transfer bytes from the inputfile to the outputfile
        byte[] buffer = new byte[1024];
        int length;
        while ((length = myInput.read(buffer)) > 0) {
            myOutput.write(buffer, 0, length);
        }

        //Close the streams
        myOutput.flush();
        myOutput.close();
        myInput.close();

    }

    void openDataBase() throws SQLException {

        //Open the database
        String myPath = DATABASE_PATH + DATABASE_NAME;
        myDataBase = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY);

    }

    @Override
    public synchronized void close() {

        if (myDataBase != null)
            myDataBase.close();

        super.close();

    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {

    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int oldversion, int newversion) {

    }

    void searchData(String vendorName, String componentName) {
        myDataBase = this.getReadableDatabase();
        Cursor cursor;
        String QueryString = "select * from " + DATABASE_TABLE + " where " + vendorName + "=" +'"'+componentName.toUpperCase()+'"';
        cursor = myDataBase.rawQuery(QueryString, null);
        if (cursor != null && cursor.moveToFirst()) {

            int s = 0;
            while (cursor.isLast() && s != 29) {
                componentNames[s] = cursor.getString(s);
                System.out.println(cursor.getString(s));
                s += 1;
            }
        }
        StringBuilder builder = new StringBuilder();
        for(String s : componentNames) {
            builder.append(s);
        }

        System.out.println(componentNames[1]);
        if (cursor != null) {
            cursor.close();
        }
    }
}

File: app/src/main/java/com/cdiamon/autocolorist/tables/DBHelper.kt
package com.cdiamon.autocolorist.tables

import android.content.Context
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import java.io.FileOutputStream
import java.io.IOException

/**
 * Created by Dmitriy on 28.12.2016.
 */
internal class DBHelper

/**
 * Constructor
 * Takes and keeps a reference of the passed context in order to access to the application assets and resources.
 *
 * @param context
 */
(private val myContext: Context) : SQLiteOpenHelper(myContext, DATABASE_NAME, null, DATABASE_VERSION) {

    private var myDataBase: SQLiteDatabase? = null

    override fun onOpen(db: SQLiteDatabase?) {
        super.onOpen(db)
        //this fixes android 9.0 problem
        db?.disableWriteAheadLogging()
    }

    @Throws(IOException::class)
    fun createDataBase() {

        val dbExist = checkDataBase()

        if (!dbExist) {

            this.readableDatabase

            copyDataBase()
        }
    }

    /**
     * Check if the database already exist to avoid re-copying the file each time you open the application.
     *
     * @return true if it exists, false if it doesn't
     */
    private fun checkDataBase(): Boolean {

        var checkDB: SQLiteDatabase? = null

        try {
            val myPath = DATABASE_PATH + DATABASE_NAME
            checkDB = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY)
        } catch (e: SQLiteException) {
            e.printStackTrace()
        }

        if (checkDB != null) {
            checkDB.close()
        }

        return checkDB != null
    }

    /**
     * Copies database from local assets-folder to the just created empty database in the
     * system folder, from where it can be accessed and handled.
     */
    private fun copyDataBase() {
        try {
            val myInput = myContext.assets.open(DATABASE_NAME)
            val outFileName = DATABASE_PATH + DATABASE_NAME
            println(outFileName)
            val myOutput = FileOutputStream(outFileName)
            myInput.copyTo(myOutput, 1024)
            myOutput.flush()
            myOutput.close()
            myInput.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Throws(SQLException::class)
    fun openDataBase() {

        val myPath = DATABASE_PATH + DATABASE_NAME
        myDataBase = SQLiteDatabase.openDatabase(myPath, null, SQLiteDatabase.OPEN_READONLY)

    }

    @Synchronized
    override fun close() {

        if (myDataBase != null) {
            myDataBase!!.close()
        }

        super.close()

    }

    override fun onCreate(sqLiteDatabase: SQLiteDatabase) {}

    override fun onUpgrade(sqLiteDatabase: SQLiteDatabase, oldversion: Int, newversion: Int) {}

    fun searchData(vendorName: String, componentName: String): Array<String?> {

        val componentNames = arrayOfNulls<String>(30)

        myDataBase = this.readableDatabase
        val cursor: Cursor?
        val QueryString = "select * from " + DATABASE_TABLE + " where " + vendorName + "=" + '"'.toString() + componentName.toUpperCase() + '"'.toString() + "LIMIT 1"
        cursor = myDataBase!!.rawQuery(QueryString, null)
        if (cursor != null && cursor.moveToFirst()) {

            var s = 0
            while (cursor.isLast && s != 29) {
                componentNames[s] = cursor.getString(s)
                println(cursor.getString(s))
                s += 1
            }
        }
        val builder = StringBuilder()
        for (s in componentNames) {
            builder.append(s)
        }

        println(componentNames[1])
        cursor?.close()

        return componentNames
    }

    companion object {


        private val DATABASE_PATH = "/data/data/com.cdiamon.autocolorist/databases/"
        private val DATABASE_NAME = "converttables.db"
        private val DATABASE_TABLE = "paints_name"
        private val DATABASE_VERSION = 1
    }
}
